src/node_jsvmapi.cc

```diff
@@ -585,96 +623,235 @@ void napi_set_return_value(napi_env e,
-  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(o)->ToObject();
 
-  return obj->Has(i);
+  if (!has) {
+    return napi_set_last_error_code(napi_invalid_arg);
+  }
+
+  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  auto obj_maybe = v8impl::V8LocalValueFromJsValue(o)->ToObject(context);
+
+  if (obj_maybe.IsEmpty()) {
+    return napi_set_last_error_code(napi_object_expected);
+  }
+
+  v8::Maybe<bool> has_maybe = obj_maybe.ToLocalChecked()->Has(context, i);
+
+  if (has_maybe.IsNothing()) {
+    return napi_set_last_error_code(napi_generic_failure);
+  }
+
+  *has = has_maybe.FromMaybe(false);
+  return napi_ok;
 }
 
-napi_value napi_get_element(napi_env e, napi_value o, uint32_t i) {
+napi_errorvalue napi_get_element(napi_env e, 
+                                 napi_value o, 
+                                 uint32_t i, 
+                                 napi_value* result) {
   NAPI_PREAMBLE(e);
-  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(o)->ToObject();
-  v8::Local<v8::Value> val = obj->Get(i);
-  // This implementation is missing a lot of details, notably error
-  // handling on invalid inputs and regarding what happens in the
-  // Set operation (error thrown, key is invalid, the bool return
-  // value of Set)
-  return v8impl::JsValueFromV8LocalValue(val);
+
+  if (!result) {
+    return napi_set_last_error_code(napi_invalid_arg);
+  }
+
+  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  auto obj_maybe = v8impl::V8LocalValueFromJsValue(o)->ToObject(context);
+
+  if (obj_maybe.IsEmpty()) {
+    return napi_set_last_error_code(napi_object_expected);
+  }
+
+  auto get_maybe = obj_maybe.ToLocalChecked()->Get(context, i);
+
+  if (get_maybe.IsEmpty()) {
+    return napi_set_last_error_code(napi_generic_failure);
+  }
+
+  *result = v8impl::JsValueFromV8LocalValue(get_maybe.ToLocalChecked());
+  return napi_ok;
 }
 
-void napi_define_property(napi_env e, napi_value o,
+napi_errorvalue napi_define_property(napi_env e, napi_value o,
     napi_property_descriptor* p) {
+  NAPI_PREAMBLE(e);
+
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
   v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(o)->ToObject();
-  v8::Local<v8::Name> name =
-    v8::String::NewFromUtf8(isolate, p->utf8name,
-      v8::NewStringType::kInternalized).ToLocalChecked();
+  auto obj_maybe = v8impl::V8LocalValueFromJsValue(o)->ToObject(context);
+
+  if (obj_maybe.IsEmpty()) {
+    return napi_set_last_error_code(napi_object_expected);
+  }
+
+  auto obj = obj_maybe.ToLocalChecked();
+
+  auto name_maybe = v8::String::NewFromUtf8(isolate, p->utf8name,
+    v8::NewStringType::kInternalized);
+
+  if (name_maybe.IsEmpty())
+  {
+    return napi_set_last_error_code(napi_generic_failure);
+  }
```

Comment:
I wonder if we should define a `CHECK_NOT_EMPTY` macro or something like that to streamline this repeated pattern.